### PR TITLE
Added fromIterator versions that capture exceptions.

### LINKED ISSUE #2334: Add fromEffectfulIterator to Stream
```scala
  final def fromEffectfulIterator[R, A]
  (iterator: ZManaged[R, Throwable, Iterator[A]]
  ): StreamEffect[R, Throwable, A] =
    StreamEffect {
      iterator.flatMap { iterator =>
        Managed.fromEffect(ZIO.runtime).flatMap { rt =>
          Managed.effectTotal { () =>
            val hasNext: Boolean = try iterator.hasNext catch {
              case e: Throwable if !rt.platform.fatal(e) =>
                StreamEffect.fail(e)
            }

            if (hasNext) {
              try iterator.next() catch {
                case e: Throwable if !rt.platform.fatal(e) => 
                  StreamEffect.fail(e)
              }
            } else StreamEffect.end
          }
        }
      }
    }
```

Allows handling exceptions that arise during iterator's `hasNext` and `next` methods.

The existing method 
```
final def fromIteratorManaged[R, E, A](iterator: ZManaged[R, E, Iterator[A]]): ZStream[R, E, A] =
    StreamEffect.fromIterator(iterator)
```
returns `Stream[E, A]` for any `E`, which means it must die if the underlying iterator throws.

### PR DESCRIPTION
**Pull request by alexknvl**
Closes #2334.

### CODE REVIEW

File: streams/jvm/src/main/scala/zio/stream/platform.scala [hunk 1]
```diff
@@ -33,27 +33,51 @@ trait ZStreamPlatformSpecificConstructors {
...
-  final def fromJavaStream[R, E, A](stream: ZIO[R, E, ju.stream.Stream[A]]): ZStream[R, E, A] =
-    ZStream.fromJavaIterator(stream.flatMap(s => UIO(s.iterator())))
+  final def fromJavaStreamTotal[A](stream: => ju.stream.Stream[A]): ZStream[Any, Nothing, A] =
+    ZStream.fromJavaIteratorTotal(stream.iterator())
+
+  /**
+   * Creates a stream from a Java stream
+   */
+  final def fromJavaStream[A](stream: => ju.stream.Stream[A]): ZStream[Any, Throwable, A] =
+    ZStream.fromJavaIteratorTotal(stream.iterator())
```
Comment (iravid) [inline streams/jvm/src/main/scala/zio/stream/platform.scala]:
```suggestion
    ZStream.fromJavaIterator(stream.iterator())
```

File: streams/jvm/src/main/scala/zio/stream/platform.scala [hunk 2]
```diff
@@ -33,27 +33,51 @@ trait ZStreamPlatformSpecificConstructors {
...
+
+  /**
+   * Creates a stream from a Java stream
+   */
+  final def fromJavaStreamEffect[R, A](stream: ZIO[R, Throwable, ju.stream.Stream[A]]): ZStream[R, Throwable, A] =
+    ZStream.fromJavaIteratorEffect(stream.flatMap(s => UIO.succeedNow(s.iterator())))
```
Comment (iravid) [inline streams/jvm/src/main/scala/zio/stream/platform.scala]:
```suggestion
    ZStream.fromJavaIteratorEffect(stream.flatMap(s => UIO(s.iterator())))
```

**Comment by alexknvl** [comment]
I'll review it myself as well, I think there are some simplifications that could be made like `.flatMap(s => UIO(...))` => `map(s => ...)`.

**Comment by iravid** [comment]
I think the call to Stream#iterator is rightly suspended as who knows what throws and what doesn’t 🤷🏼‍♂️

File: streams/shared/src/main/scala/zio/stream/StreamEffect.scala [hunk 3]
```diff
@@ -398,7 +398,59 @@ private[stream] object StreamEffect extends Serializable {
...
+                     ex = null
+                     StreamEffect.failure(ex0)
+                   } else StreamEffect.end
+                 }
+               }
+        } yield it
+      }
+    }
+
+  def fromJavaIteratorTotal[A](iterator: ju.Iterator[A]): StreamEffect[Any, Nothing, A] = {
```
Comment (alexknvl) [inline streams/shared/src/main/scala/zio/stream/StreamEffect.scala]:
```suggestion
  def fromJavaIteratorTotal[A](iterator: => ju.Iterator[A]): StreamEffect[Any, Nothing, A] = {
```

**Comment by alexknvl** [comment]
Okay, fixed one more minor problem in `StreamEffect.fromJavaIterator`.

**Comment by iravid** [comment]
Needs a format and should be good to go!

**Comment by iravid** [comment]
Awesome work. Thanks again @alexknvl.

**Pull request merged**